File: python/src/basecamp/services/schedules.py
"""Schedules service with merge-safe ``update_entry`` and read-modify-write ``edit_entry``.

``PUT /{accountId}/schedule_entries/{entryId}`` is a full replace: BC3's
``Schedules::EntriesController#update`` rebuilds the recordable from the
submitted params, so a writable field the body omits is cleared. Omitting
``summary`` leaves the entry reading back as ``"Untitled"``; omitting
``description`` clears it; omitting ``all_day`` turns an all-day event into a
midnight-to-midnight timed one. None of those is a 422 — each is a ``200`` that
quietly rewrites the entry.

**Except three.** BC3 seeds ``participant_ids``, ``url`` and ``highlighted``
from the existing recordable when the request does not address them, so for
those three an omission preserves rather than clears. That splits the writable
set in two, and the split is the whole point of this module:

* **full state** — ``summary``, ``starts_at``, ``ends_at``, ``description``,
  ``all_day``: always sent, empties included. On a full-replace endpoint ``""``
  is how a clear is expressed, never JSON null and never omission.
* **addressed-only** — ``participant_ids``, ``url``, ``highlighted``,
  ``notify``: sent only when the caller addressed them, and never seeded from
  the read-back. Resending what BC3 already preserves is redundant at best and
  wrong if the read raced a concurrent change. ``notify`` is addressed-only for
  a different reason: it is a directive, not state — sending it makes BC3
  recompute a drafted entry's subscriber list.

An explicitly empty value in the second class is an address, not an absence:
``participant_ids=[]`` clears participants, ``url=""`` clears the join link,
``highlighted=False`` removes the highlight. Each survives body compaction,
which strips only ``None``.

The response spells the join link ``join_url`` — the entry's ``url`` is its own
Basecamp API URL, written by the recordings partial that renders first. Echoing
the response's ``url`` into the request's ``url`` would therefore write the
entry's API URL into its join link, which is why the read-side seed reads
``join_url`` and the composite never seeds it at all.

Both composites compose the public ``get_entry`` and ``replace_entry`` methods,
so hooks observe the two wire operations (``get_entry`` then ``replace_entry``),
not a synthetic composite.

**Non-recurring entries only.** BC3's ``ensure_non_recurring_event`` 302-redirects
both ``show`` and ``update`` for a recurring entry, so this route serves
non-recurring entries. A recurring entry surfaces as a redirect the SDK does not
follow for the PUT, or as an unexpected body on the GET — which the read guards
below refuse before anything is written.

Neither composite is atomic: there is no conditional-update signal on this
endpoint, so a concurrent write between the GET and PUT is overwritten — last
write wins for the whole representation. The window is one round-trip. Use
``replace_entry`` to overwrite deliberately.
"""

from __future__ import annotations

from typing import Any

from basecamp.errors import UsageError
from basecamp.generated.services.schedules import (
    AsyncSchedulesService as _GeneratedAsyncSchedulesService,
)
from basecamp.generated.services.schedules import SchedulesService as _GeneratedSchedulesService
from basecamp.services._merge_safe import (
    require_mapping,
    required_writable_boolean,
    required_writable_string,
    writable_boolean,
    writable_id_list,
    writable_string,
)

_ESCAPE = "replace_entry()"
_RECORD = "ScheduleEntry"

#: Resent verbatim on every PUT, empties included — BC3 clears what the body omits.
_FULL_STATE = ("summary", "starts_at", "ends_at", "description", "all_day")

#: Sent only when the caller addressed them. BC3's ``preservedOnOmission`` carve-out
#: covers the first three; ``notify`` is a directive rather than entry state.
_CARVE_OUTS = frozenset({"participant_ids", "url", "highlighted", "notify"})


def _require_entry(entry: object) -> dict[str, Any]:
    return require_mapping(entry, record=_RECORD, operation="GetScheduleEntry", escape=_ESCAPE)


def _fields_from_entry(body: dict[str, Any]) -> dict[str, Any]:
    """Derive an entry's full writable state from a GET response.

    Every value here is resent in the full-replace PUT, so every value is
    validated before it is read. A plain ``or ""`` would coerce each falsey
    non-string to ``""`` — erasing the field on a call that never mentioned it —
    and pass ``42``/``True`` straight through to be written verbatim. Python has
    no typed decoder between the GET and this read (``get_entry`` returns
    ``dict[str, Any]``), so the check is explicit work here rather than
    something the layer below already did. See
    :mod:`basecamp.services._merge_safe` and #576.

    The five fields read differently because the spec models them differently:

    * ``summary`` is ``@required`` and ``Schedule::Entry#summary`` is
      ``super.presence || "Untitled"``, so absent, null or blank is a malformed
      response rather than an empty summary;
    * ``starts_at``/``ends_at`` are ``@required`` and ``NOT NULL`` — and their
      wire value is a bare date (``"2026-06-05"``) for an all-day entry and a
      timestamp otherwise, so they are round-tripped verbatim and never parsed;
    * ``all_day`` is ``@required``, ``NOT NULL DEFAULT false``, and the one
      field whose guard cannot be a truthiness test: ``False`` is the value it
      most needs to admit, and defaulting a missing one to ``False`` would
      convert an all-day event into a midnight-to-midnight timed one;
    * ``description`` is optional and nullable, so absent or null is a genuinely
      empty description.
    """
    return {
        "summary": required_writable_string(body, "summary", record=_RECORD, escape=_ESCAPE),
        "starts_at": required_writable_string(body, "starts_at", record=_RECORD, escape=_ESCAPE),
        "ends_at": required_writable_string(body, "ends_at", record=_RECORD, escape=_ESCAPE),
        "description": writable_string(body, "description", record=_RECORD, escape=_ESCAPE),
        "all_day": required_writable_boolean(body, "all_day", record=_RECORD, escape=_ESCAPE),
    }


def _carve_out_seeds(body: dict[str, Any]) -> dict[str, Any]:
    """Seed the carve-outs **for reading inside an edit block**, never for sending.

    :meth:`SchedulesService.edit_entry` exposes the current join link, highlight
    and participants so a block can inspect them before deciding; only an
    assignment puts one on the wire. The values are still guarded, because
    ``entry.url = entry.url`` is a legitimate write and would otherwise carry a
    malformed value into the PUT.

    ``url`` is read from the response's ``join_url``: the entry's own ``url`` is
    its Basecamp API URL. ``participant_ids`` is projected from the
    ``participants`` array. ``notify`` has no read-back at all — it is a send
    directive — so it starts ``False`` and only reaches the wire if the block
    assigns it.
    """
    return {
        "participant_ids": writable_id_list(body, "participants", record=_RECORD, escape=_ESCAPE),
        "url": writable_string(body, "join_url", record=_RECORD, escape=_ESCAPE),
        "highlighted": writable_boolean(body, "highlighted", record=_RECORD, escape=_ESCAPE),
        "notify": False,
    }


def _replace_kwargs(fields: dict[str, Any], addressed: dict[str, Any]) -> dict[str, Any]:
    """Serialize full writable state plus whatever the caller addressed.

    The five full-state fields are always named, empties included. The
    addressed-only keys are merged in exactly as passed, so an explicit ``""``,
    ``False`` or ``[]`` reaches the wire while an unaddressed key never appears
    — the transport's compaction strips ``None`` only, so absence has to be
    expressed by leaving the key out here.
    """
    return {
        "summary": fields["summary"],
        "starts_at": fields["starts_at"],
        "ends_at": fields["ends_at"],
        "description": fields["description"],
        "all_day": fields["all_day"],
        **addressed,
    }


def _overlay(fields: dict[str, Any], **updates: Any) -> dict[str, Any]:
    for key, value in updates.items():
        if value is not None:
            fields[key] = value
    return fields


def _addressed_carve_outs(**candidates: Any) -> dict[str, Any]:
    """Keep the carve-outs the caller actually passed.

    ``None`` is the only "not addressed" signal, so ``""``, ``False`` and ``[]``
    all survive — each of them is an instruction to clear, and handing a clear
    back to BC3's carve-out would preserve instead.
    """
    return {key: value for key, value in candidates.items() if value is not None}


class _ScheduleEntryEditBase:
    """Shared writable state for :class:`ScheduleEntryEdit` / :class:`AsyncScheduleEntryEdit`.

    Inside the ``with`` block the edit object exposes the entry's full writable
    state — ``summary``, ``starts_at``, ``ends_at``, ``description``,
    ``all_day`` — which is always written back, plus the four addressed-only
    fields ``participant_ids``, ``url``, ``highlighted`` and ``notify``, which
    are written back **only if the block assigns them**.

    That dirty set is keyed on setter invocation, not on value comparison. A
    snapshot diff would drop ``entry.url = entry.url``, and the intent behind
    that statement is not recoverable from the value: re-asserting the join link
    the read returned is a write, and a caller who wrote it deserves to have it
    sent. Clearing a field means setting it empty (``""``, ``[]``, ``False``);
    an untouched full-state field keeps its current value, and an untouched
    carve-out is left to BC3, which preserves it.
    """

    summary: str
    starts_at: str
    ends_at: str
    description: str
    all_day: bool
    participant_ids: list[int]
    url: str
    highlighted: bool
    notify: bool

    def __init__(self, entry_id: int) -> None:
        # First, and before any tracked attribute can be set: __setattr__ reads it.
        self._touched: set[str] = set()
        self._entry_id = entry_id
        self._result: dict[str, Any] | None = None
        self._completed = False

    def __setattr__(self, name: str, value: Any) -> None:
        """Record an assignment to a carve-out, then store it.

        This — not a comparison against the read-back — is what puts
        ``participant_ids``, ``url``, ``highlighted`` or ``notify`` on the wire.
        """
        if name in _CARVE_OUTS:
            self._touched.add(name)
        object.__setattr__(self, name, value)

    def _load(self, entry: dict[str, Any]) -> None:
        body = _require_entry(entry)
        for key, value in _fields_from_entry(body).items():
            setattr(self, key, value)
        # Seeded for reading only. Going through object.__setattr__ keeps the
        # seeds out of the dirty set, so a block that merely reads `url` does
        # not write it back.
        for key, value in _carve_out_seeds(body).items():
            object.__setattr__(self, key, value)

    def _fields(self) -> dict[str, Any]:
        return {key: getattr(self, key) for key in _FULL_STATE}

    def _addressed(self) -> dict[str, Any]:
        addressed = {key: getattr(self, key) for key in sorted(self._touched)}
        for key, value in addressed.items():
            if value is None:
                raise UsageError(
                    f"{key} was set to None; a full write has no None state — "
                    f'use "" or [] or False to clear it, or leave it unassigned to keep it'
                )
        return addressed

    @property
    def result(self) -> dict[str, Any]:
        """The updated entry, available after the ``with`` block exits cleanly."""
        if not self._completed:
            raise RuntimeError("edit has not completed")
        assert self._result is not None
        return self._result


class ScheduleEntryEdit(_ScheduleEntryEditBase):
    """Read-modify-write context manager returned by :meth:`SchedulesService.edit_entry`.

    Entering the block GETs the current entry; exiting cleanly PUTs the whole
    representation back. If the block raises, the edit aborts and nothing is
    written.
    """

    def __init__(self, service: SchedulesService, entry_id: int) -> None:
        super().__init__(entry_id)
        self._service = service

    def __enter__(self) -> ScheduleEntryEdit:
        self._load(self._service.get_entry(entry_id=self._entry_id))
        return self

    def __exit__(self, exc_type: object, exc: object, tb: object) -> None:
        if exc_type is None:
            self._result = self._service.replace_entry(
                entry_id=self._entry_id, **_replace_kwargs(self._fields(), self._addressed())
            )
            self._completed = True


class AsyncScheduleEntryEdit(_ScheduleEntryEditBase):
    """Async twin of :class:`ScheduleEntryEdit`, for ``async with``."""

    def __init__(self, service: AsyncSchedulesService, entry_id: int) -> None:
        super().__init__(entry_id)
        self._service = service

    async def __aenter__(self) -> AsyncScheduleEntryEdit:
        self._load(await self._service.get_entry(entry_id=self._entry_id))
        return self

    async def __aexit__(self, exc_type: object, exc: object, tb: object) -> None:
        if exc_type is None:
            self._result = await self._service.replace_entry(
                entry_id=self._entry_id, **_replace_kwargs(self._fields(), self._addressed())
            )
            self._completed = True


class SchedulesService(_GeneratedSchedulesService):
    """Schedules service with merge-safe ``update_entry`` and ``edit_entry`` on top
    of the generated surface (``get_entry``, ``replace_entry``, ...)."""

    def update_entry(
        self,
        *,
        entry_id: int,
        summary: str | None = None,
        starts_at: str | None = None,
        ends_at: str | None = None,
        description: str | None = None,
        all_day: bool | None = None,
        participant_ids: list[int] | None = None,
        url: str | None = None,
        highlighted: bool | None = None,
        notify: bool | None = None,
    ) -> dict[str, Any]:
        """Set the given fields on a schedule entry and preserve everything else.

        GETs the current entry, overlays the explicitly-passed keyword
        arguments, and PUTs the full representation back. An omitted (``None``)
        full-state field is untouched, guaranteed; an explicitly-passed ``""``
        or ``False`` sets it.

        ``participant_ids``, ``url``, ``highlighted`` and ``notify`` are sent
        only when passed and are never read back from the entry: BC3 preserves
        the first three when the request does not address them, and ``notify``
        is a send directive rather than entry state. Passing ``[]``, ``""`` or
        ``False`` for one of them is an address, and clears.

        ``url`` is the join link — the request spelling of the field the
        response returns as ``join_url``.

        Non-recurring entries only: BC3 302-redirects both hops for a recurring
        entry. Not atomic — a concurrent write between the GET and PUT is
        overwritten (last write wins for the whole representation; the window is
        one round-trip). Use :meth:`replace_entry` to overwrite deliberately.
        """
        fields = _overlay(
            _fields_from_entry(_require_entry(self.get_entry(entry_id=entry_id))),
            summary=summary,
            starts_at=starts_at,
            ends_at=ends_at,
            description=description,
            all_day=all_day,
        )
        addressed = _addressed_carve_outs(
            participant_ids=participant_ids, url=url, highlighted=highlighted, notify=notify
        )
        return self.replace_entry(entry_id=entry_id, **_replace_kwargs(fields, addressed))

    def edit_entry(self, *, entry_id: int) -> ScheduleEntryEdit:
        """Open a read-modify-write edit of a schedule entry, as a context manager.

        Entering the ``with`` block GETs the current entry and exposes its full
        writable state; exiting cleanly PUTs the whole representation back.
        Clearing a field means setting it empty (``""``, ``[]``, ``False``). If
        the block raises, nothing is written. The updated entry is available as
        ``.result`` after the block::

            with client.schedules.edit_entry(entry_id=123) as e:
                e.summary = f"🚨 {e.summary}"
                e.description = ""  # clearing = setting empty on a full object
                e.url = e.url  # re-asserting the join link IS a write

        ``participant_ids``, ``url``, ``highlighted`` and ``notify`` reach the
        wire only if the block assigns them — reading one does not. The last
        line above therefore sends the join link, even though its value did not
        change: the dirty set is keyed on the assignment, not on a comparison.

        Non-recurring entries only. Not atomic: a concurrent write between the
        GET and PUT is overwritten (last write wins for the whole
        representation; the window is one round-trip).
        """
        return ScheduleEntryEdit(self, entry_id)


class AsyncSchedulesService(_GeneratedAsyncSchedulesService):
    """Async schedules service with merge-safe ``update_entry`` and ``edit_entry``."""

    async def update_entry(
        self,
        *,
        entry_id: int,
        summary: str | None = None,
        starts_at: str | None = None,
        ends_at: str | None = None,
        description: str | None = None,
        all_day: bool | None = None,
        participant_ids: list[int] | None = None,
        url: str | None = None,
        highlighted: bool | None = None,
        notify: bool | None = None,
    ) -> dict[str, Any]:
        """Async twin of :meth:`SchedulesService.update_entry`."""
        fields = _overlay(
            _fields_from_entry(_require_entry(await self.get_entry(entry_id=entry_id))),
            summary=summary,
            starts_at=starts_at,
            ends_at=ends_at,
            description=description,
            all_day=all_day,
        )
        addressed = _addressed_carve_outs(
            participant_ids=participant_ids, url=url, highlighted=highlighted, notify=notify
        )
        return await self.replace_entry(entry_id=entry_id, **_replace_kwargs(fields, addressed))

    def edit_entry(self, *, entry_id: int) -> AsyncScheduleEntryEdit:
        """Async twin of :meth:`SchedulesService.edit_entry`, for ``async with``::

        async with client.schedules.edit_entry(entry_id=123) as e:
            e.summary = f"🚨 {e.summary}"
        updated = e.result
        """
        return AsyncScheduleEntryEdit(self, entry_id)

File: rust/basecamp-sdk/src/services/schedules.rs
//! Schedules: the generated wire methods plus the merge-safe `update_entry` and
//! read-modify-write `edit_entry` of SPEC §5 "Merge-Safe Write Surface (Schedule
//! Entries)".
//!
//! `PUT /schedule_entries/{entryId}` is a full replace with declared carve-outs. BC3
//! rebuilds the recordable from the permitted params, so a body that omits `description`
//! erases it, one that omits `summary` leaves the entry reading back as "Untitled", and
//! one that omits `all_day` turns an all-day entry into a midnight-to-midnight timed one.
//! Three writable fields are exempt: `participant_ids`, `url` and `highlighted` are
//! preserved server-side when the body does not name them.
//!
//! The composites therefore split the writable set in two. The five replaced fields —
//! `summary`, `description`, `all_day`, `starts_at`, `ends_at` — are read back and
//! resent whether or not the caller touched them. The three carve-outs, and the `notify`
//! directive, reach the wire only when the caller addressed them, and then apply
//! normally: `vec![]` clears the participants, `""` clears the join link, `false` removes
//! the highlight. Resending an unaddressed carve-out would be redundant at best and wrong
//! if the read raced a change — and the response spells the join link `join_url`, because
//! the entry's own `url` is its Basecamp API URL.
//!
//! Both composites go through [`SchedulesService::get_entry`] and
//! [`SchedulesService::replace_entry`], so hooks observe `GetScheduleEntry` then
//! `ReplaceScheduleEntry` under their own identities. Recurring entries are unreachable
//! here: BC3 redirects both reads and writes of one to its occurrence.
//!
//! Neither composite is atomic: a concurrent write between the GET and the PUT is
//! overwritten, last write wins, with a window of one round-trip. Use `replace_entry` to
//! overwrite deliberately.

pub use crate::generated::services::schedules::{ListScheduleEntriesParams, SchedulesService};

use crate::error::Error;
use crate::generated::types::{ReplaceScheduleEntryRequestContent, ScheduleEntry};
use crate::types::FlexibleTime;

/// The fields a merge-safe [`SchedulesService::update_entry`] may set. A `None` member is
/// not addressed: for the five replaced fields the read-back value is resent, and for
/// the carve-outs the key stays off the wire so BC3 keeps what it holds. A `Some` member
/// is written, so `Some(String::new())`, `Some(vec![])` and `Some(false)` are clears.
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct UpdateScheduleEntryRequest {
    /// Plain-text summary. Cleared, the entry reads back as "Untitled".
    pub summary: Option<String>,
    /// Rich text description (HTML).
    pub description: Option<String>,
    /// Whether the entry spans whole days.
    pub all_day: Option<bool>,
    /// The start, a bare date for an all-day entry or a full timestamp otherwise, sent
    /// verbatim.
    pub starts_at: Option<FlexibleTime>,
    /// The end. See `starts_at` for the date-versus-timestamp rule.
    pub ends_at: Option<FlexibleTime>,
    /// Replaces the participants; `Some(vec![])` clears them.
    pub participant_ids: Option<Vec<i64>>,
    /// The join link; `Some(String::new())` clears it.
    pub url: Option<String>,
    /// Whether the entry is highlighted; `Some(false)` removes the highlight.
    pub highlighted: Option<bool>,
    /// Whether to notify participants. A directive, not entry state.
    pub notify: Option<bool>,
}

/// A schedule entry's writable state, handed to the [`SchedulesService::edit_entry`]
/// closure.
///
/// The five replaced fields are plain members: seeded from the read-back and resent
/// whether or not the closure touches them, so clearing one means setting it `""`. The
/// carve-outs are seeded for reading — [`url`](Self::url) from the response's `join_url`,
/// [`participant_ids`](Self::participant_ids) projected from its `participants`,
/// [`highlighted`](Self::highlighted) as returned — but reach the wire only when the
/// matching setter is invoked. Dirty tracking is by setter invocation, not value
/// comparison: assigning the value the read returned is still an address and is sent.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ScheduleEntryFields {
    /// Plain-text summary. Set `""` to clear; the entry then reads back as "Untitled".
    pub summary: String,
    /// Rich text description (HTML). Set `""` to clear.
    pub description: String,
    /// Whether the entry spans whole days.
    pub all_day: bool,
    /// The start, round-tripped verbatim: a bare date for an all-day entry, a full
    /// timestamp otherwise.
    pub starts_at: FlexibleTime,
    /// The end, round-tripped verbatim.
    pub ends_at: FlexibleTime,
    participant_ids: Option<Vec<i64>>,
    url: Option<String>,
    highlighted: Option<bool>,
    notify: Option<bool>,
    addressed: Addressed,
}

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
struct Addressed {
    participant_ids: bool,
    url: bool,
    highlighted: bool,
    notify: bool,
}

impl ScheduleEntryFields {
    /// The participants' ids as the read-back projected them, or `None` when the
    /// response carried no `participants`.
    pub fn participant_ids(&self) -> Option<&[i64]> {
        self.participant_ids.as_deref()
    }

    /// The join link as the read-back's `join_url`, or `None` when the entry has none.
    pub fn url(&self) -> Option<&str> {
        self.url.as_deref()
    }

    /// Whether the entry is highlighted, or `None` when the response did not say.
    pub fn highlighted(&self) -> Option<bool> {
        self.highlighted
    }

    /// Whether participants will be notified, once [`set_notify`](Self::set_notify) has
    /// been called.
    pub fn notify(&self) -> Option<bool> {
        self.notify
    }

    /// Replaces the participants. `vec![]` removes everyone.
    pub fn set_participant_ids(&mut self, participant_ids: Vec<i64>) {
        self.participant_ids = Some(participant_ids);
        self.addressed.participant_ids = true;
    }

    /// Sets the join link. `""` removes it.
    pub fn set_url(&mut self, url: impl Into<String>) {
        self.url = Some(url.into());
        self.addressed.url = true;
    }

    /// Sets the highlight. `false` removes it.
    pub fn set_highlighted(&mut self, highlighted: bool) {
        self.highlighted = Some(highlighted);
        self.addressed.highlighted = true;
    }

    /// Asks the server to notify participants about this write.
    pub fn set_notify(&mut self, notify: bool) {
        self.notify = Some(notify);
        self.addressed.notify = true;
    }

    /// `summary`, `starts_at` and `ends_at` are required on the response and BC3 never
    /// renders them blank — `Schedule::Entry#summary` falls back to "Untitled" and the
    /// bounds are NOT NULL columns — so a blank one in a 2xx body is malformed rather than
    /// empty: resending it would blank the real value on a call that touched something
    /// else. `all_day` is taken as decoded, `false` included.
    fn from_entry(entry: &ScheduleEntry) -> Result<ScheduleEntryFields, Error> {
        require_rendered(&entry.summary, "summary")?;
        require_rendered(entry.starts_at.as_str(), "starts_at")?;
        require_rendered(entry.ends_at.as_str(), "ends_at")?;
        Ok(ScheduleEntryFields {
            summary: entry.summary.clone(),
            description: entry.description.clone().unwrap_or_default(),
            all_day: entry.all_day,
            starts_at: entry.starts_at.clone(),
            ends_at: entry.ends_at.clone(),
            participant_ids: entry
                .participants
                .as_ref()
                .map(|people| people.iter().map(|person| person.id).collect()),
            url: entry.join_url.clone(),
            highlighted: entry.highlighted,
            notify: None,
            addressed: Addressed::default(),
        })
    }

    /// The carve-outs go through the same setters a closure would call, so "the caller
    /// addressed this" has one implementation.
    fn apply(&mut self, request: &UpdateScheduleEntryRequest) {
        if let Some(summary) = &request.summary {
            self.summary.clone_from(summary);
        }
        if let Some(description) = &request.description {
            self.description.clone_from(description);
        }
        if let Some(all_day) = request.all_day {
            self.all_day = all_day;
        }
        if let Some(starts_at) = &request.starts_at {
            self.starts_at.clone_from(starts_at);
        }
        if let Some(ends_at) = &request.ends_at {
            self.ends_at.clone_from(ends_at);
        }
        if let Some(participant_ids) = &request.participant_ids {
            self.set_participant_ids(participant_ids.clone());
        }
        if let Some(url) = &request.url {
            self.set_url(url.clone());
        }
        if let Some(highlighted) = request.highlighted {
            self.set_highlighted(highlighted);
        }
        if let Some(notify) = request.notify {
            self.set_notify(notify);
        }
    }

    /// The five replaced fields always, empties included: `""` is how a clear is spelled
    /// on a full-replace endpoint, never `null` (SPEC §18) and never omission. A carve-out
    /// is passed only when addressed; unaddressed, it is `None` and body compaction leaves
    /// no key at all.
    fn into_body(self) -> ReplaceScheduleEntryRequestContent {
        let addressed = self.addressed;
        ReplaceScheduleEntryRequestContent {
            summary: Some(self.summary),
            starts_at: self.starts_at.0,
            ends_at: self.ends_at.0,
            description: Some(self.description),
            participant_ids: addressed
                .participant_ids
                .then(|| self.participant_ids.unwrap_or_default()),
            all_day: Some(self.all_day),
            notify: addressed.notify.then_some(self.notify).flatten(),
            url: addressed.url.then(|| self.url.unwrap_or_default()),
            highlighted: addressed.highlighted.then_some(self.highlighted).flatten(),
        }
    }
}

fn require_rendered(value: &str, key: &str) -> Result<(), Error> {
    if value.trim().is_empty() {
        Err(Error::malformed_response(format!(
            "GetScheduleEntry returned a schedule entry with a blank \"{key}\", which the API \
             never renders"
        ))
        .with_hint(MALFORMED_HINT))
    } else {
        Ok(())
    }
}

const MALFORMED_HINT: &str = "The merge-safe update_entry and edit_entry resend this \
                              record's fields verbatim, so a malformed response cannot be \
                              written back safely. Use replace_entry to write the record \
                              deliberately.";

impl SchedulesService<'_> {
    /// Sets the given fields on a schedule entry and preserves the rest: GETs the current
    /// entry, resends its five replaced fields with the `Some` members of `request`
    /// overlaid, adds any carve-out the request addressed, and PUTs through
    /// [`SchedulesService::replace_entry`].
    ///
    /// Not atomic — see the module docs for the GET→PUT race.
    pub async fn update_entry(
        &self,
        entry_id: i64,
        request: &UpdateScheduleEntryRequest,
    ) -> Result<ScheduleEntry, Error> {
        let mut fields = ScheduleEntryFields::from_entry(&self.get_entry(entry_id).await?)?;
        fields.apply(request);
        self.replace_entry(entry_id, &fields.into_body()).await
    }

    /// Applies a read-modify-write closure to a schedule entry: GETs the current entry,
    /// hands the closure its writable state, and PUTs it back through
    /// [`SchedulesService::replace_entry`]. An error from the closure aborts before
    /// anything is written.
    ///
    /// ```no_run
    /// # async fn example(account: basecamp_sdk::AccountClient) -> Result<(), basecamp_sdk::Error> {
    /// account
    ///     .schedules()
    ///     .edit_entry(1069479523, |entry| {
    ///         entry.summary = format!("🚨 {}", entry.summary);
    ///         if entry.url().is_some_and(|url| url.starts_with("https://meet.example.com/")) {
    ///             entry.set_url("");
    ///         }
    ///         Ok(())
    ///     })
    ///     .await?;
    /// # Ok(())
    /// # }
    /// ```
    ///
    /// Not atomic — see the module docs for the GET→PUT race.
    pub async fn edit_entry<F>(&self, entry_id: i64, mutate: F) -> Result<ScheduleEntry, Error>
    where
        F: FnOnce(&mut ScheduleEntryFields) -> Result<(), Error>,
    {
        let mut fields = ScheduleEntryFields::from_entry(&self.get_entry(entry_id).await?)?;
        mutate(&mut fields)?;
        self.replace_entry(entry_id, &fields.into_body()).await
    }
}
